### PR TITLE
Trimmed text

### PR DESCRIPTION
**Pull request by staff0rd**
Allows `PIXI.Text` to be trimmed of transparency for better layouts.

Opt in only;
~~~javascript
new PIXI.Text('my text', null, null, true);
~~~

https://codepen.io/staff0rd/pen/KaVdXZ?editors=0010

### CODE REVIEW

File: src/core/text/Text.js [hunk 1]
```diff
@@ -103,6 +104,13 @@ export default class Text extends Sprite
         this.style = style;
 
         this.localStyleID = -1;
+
+        /**
+         * Trim transparent borders
+         *
+         * @member {boolean}
+         */
+        this.trim = trim;
```
Comment (bigtimebuddy) [inline src/core/text/Text.js]: How about `this.trim = trim || false;` or set `trim = false` as the default param to make sure defaults to an actually bool and not undefined.

**Comment by bigtimebuddy** [comment]
I think it makes sense to add `trim` as a member of TextStyle. Seems similar to the `padding` option.

**Comment by staff0rd** [comment]
Yeah - i'll move it in there if required - I wasn't sure of whether style was supposed be reflecting css style.  Would certainly be nicer pass via style however.  Will make the changes if there's a desire to merge, thought this might be too hacky.

**Comment by bigtimebuddy** [comment]
@staff0rd I hear you. Text is already super slow, adding a couple `getImageData` calls and looping through the pixels seems even more expensive. Not sure if the juice is worth the squeeze in this case. Can you describe situation where trimming would be really important?

**Comment by staff0rd** [comment]
1. `Text.texture`s take into account ascenders & descenders, even when the current text value does not use them, sometimes resulting in skewed padding and odd layout.  Most visible at larger font sizes, or with "old style" numbers (see number positions of [Raleway](https://fonts.google.com/specimen/Raleway))
2. `calculateFontProperties` is not super accurate, particularly on irregular fonts.
3.  Aligning any given text (with any given font) flush (to edges) on any given side without manually tweaking is not possible without trimming.
4. Bordering text values with equal padding for any font (see #3476)

**Comment by ivanpopelyshev** [comment]
Yeah, I dont know whether to leave it there or move into TextStyle.

**Comment by themoonrat** [comment]
In the unknown of deciding between it being in TextStyle or Text... I'll throw in a third option and say neither. Isn't this trim functionality more of a util that any sprite could use? Or a util that any generator of a texture could use?

**Comment by staff0rd** [comment]
@themoonrat I agree that the actual trim functionality could at some stage be pulled out and go somewhere more generic, however I'm trying to keep the scope small to start with to encourage merging.

**Comment by bigtimebuddy** [comment]
Love the utility approach. Seems this could be generalized and trimming is not specific to Text.

**Comment by themoonrat** [comment]
@staff0rd I understand; I just feel uneasy about yet another optional constructor parameter for Text; I was a bit meh on passing a canvas through, but this one is a tipping balance from me I'm afraid! And it almost, but doesn't quite, fit within TextStyle remit.

**Comment by bigtimebuddy** [comment]
I agree with @themoonrat on adding additional constructor args

**Comment by bigtimebuddy** [comment]
How do you want to proceed with this @staff0rd?

**Comment by staff0rd** [comment]
I've moved it into TextStyle so it looks like this now;

~~~javascript
const text = new PIXI.Text("my text", { trim: true });
~~~
Updated pen; https://codepen.io/staff0rd/pen/KaVdXZ?editors=0010

Let me know where if you want the trimming function pulled out and put elsewhere.

File: src/core/text/Text.js [hunk 2]
```diff
@@ -354,6 +363,79 @@ export default class Text extends Sprite
...
+        }
+
+        const trimHeight = bound.bottom - bound.top + 1;
+        const trimWidth = bound.right - bound.left;
+        const trimmed = this.context.getImageData(bound.left, bound.top, trimWidth, trimHeight);
+
+        return {
+            trimHeight,
+            trimWidth,
+            trimmed,
```
Comment (bigtimebuddy) [inline src/core/text/Text.js]: Could you rename these output properties to `height`, `width` and `data`? The "trim" part of the names is implied.

Comment (ivanpopelyshev) [inline src/core/text/Text.js]: +1

**Comment by staff0rd** [comment]
@bigtimebuddy applied those changes

**Pull request merged**